File: modules/brk-client/tests/tree.js
/**
 * Comprehensive test that fetches all endpoints in the tree.
 */

import { BrkClient } from "../index.js";

/**
 * Recursively collect all metric patterns from the tree.
 * @param {Record<string, any>} obj
 * @param {string} path
 * @returns {Array<{path: string, metric: Record<string, any>, indexes: string[]}>}
 */
function getAllMetrics(obj, path = "") {
  const metrics = [];

  for (const key of Object.keys(obj)) {
    if (key.startsWith("_")) continue;

    const attr = obj[key];
    if (!attr || typeof attr !== "object") continue;

    const currentPath = path ? `${path}.${key}` : key;

    // Check if this is a metric pattern (has 'by' property with index getters)
    if (attr.by && typeof attr.by === "object") {
      const indexes = Object.keys(attr.by).filter(
        (k) => !k.startsWith("_") && typeof attr.by[k] === "object",
      );
      if (indexes.length > 0) {
        metrics.push({ path: currentPath, metric: attr, indexes });
      }
    }

    // Recurse into nested tree nodes
    if (typeof attr === "object" && !Array.isArray(attr)) {
      metrics.push(...getAllMetrics(attr, currentPath));
    }
  }

  return metrics;
}

// Endpoints with sparse data (holes at the end) - skip these
const SKIP_ENDPOINTS = new Set([
  "distribution.addressesData.empty.by.emptyaddressindex",
  "distribution.addressesData.loaded.by.loadedaddressindex",
]);

async function testAllEndpoints() {
  const client = new BrkClient({ baseUrl: "http://localhost:3110", timeout: 15000 });

  const metrics = getAllMetrics(client.metrics);
  console.log(`\nFound ${metrics.length} metrics`);

  let success = 0;
  let skipped = 0;

  for (const { path, metric, indexes } of metrics) {
    for (const idxName of indexes) {
      const fullPath = `${path}.by.${idxName}`;
      if (SKIP_ENDPOINTS.has(fullPath)) {
        skipped++;
        console.log(`SKIP: ${fullPath} -> sparse data`);
        continue;
      }
      try {
        const endpoint = metric.by[idxName];
        await endpoint.last(0);
        success++;
        console.log(`OK: ${fullPath}`);
      } catch (e) {
        console.log(
          `FAIL: ${fullPath} -> ${e instanceof Error ? e.message : e}`,
        );
        return;
      }
    }
  }

  console.log(`\n=== Results ===`);
  console.log(`Success: ${success}`);
  console.log(`Skipped: ${skipped}`);
}

testAllEndpoints();

File: packages/brk_client/tests/tree.py
"""Comprehensive test that fetches all endpoints in the tree."""

from brk_client import BrkClient


def get_all_metrics(obj, path=""):
    """Recursively collect all MetricPattern instances from the tree."""
    metrics = []

    for attr_name in dir(obj):
        if attr_name.startswith("_"):
            continue

        try:
            attr = getattr(obj, attr_name)
        except Exception:
            continue

        current_path = f"{path}.{attr_name}" if path else attr_name

        # Check if this is a metric pattern (has 'by' attribute with index methods)
        if hasattr(attr, "by"):
            by = attr.by
            indexes = []
            for idx_name in dir(by):
                if not idx_name.startswith("_") and callable(
                    getattr(by, idx_name, None)
                ):
                    indexes.append(idx_name)
            if indexes:
                metrics.append((current_path, attr, indexes))

        # Recurse into nested tree nodes
        if hasattr(attr, "__dict__") and not callable(attr):
            metrics.extend(get_all_metrics(attr, current_path))

    return metrics


def test_all_endpoints():
    """Test fetching last 3 values from all metric endpoints."""
    client = BrkClient("http://localhost:3110")

    metrics = get_all_metrics(client.metrics)
    print(f"\nFound {len(metrics)} metrics")

    success = 0
    failed = 0
    errors = []

    for path, metric, indexes in metrics:
        for idx_name in indexes:
            try:
                by = metric.by
                endpoint = getattr(by, idx_name)()
                # Use the new idiomatic API: tail(3).fetch() or [-3:].fetch()
                res = endpoint.tail(3).fetch()
                count = len(res["data"])
                if count != 3:
                    failed += 1
                    error_msg = (
                        f"FAIL: {path}.by.{idx_name}() -> expected 3, got {count}"
                    )
                    errors.append(error_msg)
                    print(error_msg)
                else:
                    success += 1
                    print(f"OK: {path}.by.{idx_name}() -> {count} items")
            except Exception as e:
                failed += 1
                error_msg = f"FAIL: {path}.by.{idx_name}() -> {e}"
                errors.append(error_msg)
                print(error_msg)

    print("\n=== Results ===")
    print(f"Success: {success}")
    print(f"Failed: {failed}")

    if errors:
        print("\nErrors:")
        for err in errors[:10]:  # Show first 10 errors
            print(f"  {err}")
        if len(errors) > 10:
            print(f"  ... and {len(errors) - 10} more")

    assert failed == 0, f"{failed} endpoints failed"


if __name__ == "__main__":
    test_all_endpoints()
